liquidctl-completion.sh

```diff
@@ -0,0 +1,273 @@
+            return
+            ;;
+        --pump-mode)
+            COMPREPLY=($(compgen -W "balanced quiet extreme" -- "$cur"))
+            return
+            ;;
+        esac
+
+
+    local i=1 cmd
+
+  # find the subcommand - first word after the flags
+  while [[ "$i" -lt "$COMP_CWORD" ]]
+  do
+      local s="${COMP_WORDS[i]}"
+      case "$s" in
+          --help | --version)
+              COMPREPLY=()
+              return
+              ;;
+          -*) ;;
+          initialize | list | status | set )
+              cmd="$s"
+              break
+              ;;
+      esac
+      (( i++ ))
+  done
+
+  if [[ "$i" -eq "$COMP_CWORD" ]]
+  then
+
+      COMPREPLY=($(compgen -W "$commands $options_with_args $boolean_options" -- "$cur"))
+      return # return early if we're still completing the 'current' command
+  fi
+
+  # we've completed the 'current' command and now need to call the next completion function
+  # subcommands have their own completion functions
+  case "$cmd" in
+      list) COMPREPLY=""  ;;
+      initialize) _liquidctl_initialize_command ;;
+      status) COMPREPLY="" ;;
+      set) _liquidctl_set_command ;;
+      *)          ;;
+  esac
+}
+
+_liquidctl_initialize_command ()
+{  
+    local i=1 subcommand_index
+
+  # find the sub command (either a fan or an led to set)
+  while [[ $i -lt $COMP_CWORD ]]; do
+      local s="${COMP_WORDS[i]}"
+      case "$s" in
+          all)
+              subcommand_index=$i
+              break
+              ;;
+      esac
+      (( i++ ))
+  done
+
+
+  if [[ "$i" -eq "$COMP_CWORD" ]]
+  then
+      local cur="${COMP_WORDS[COMP_CWORD]}"
+      COMPREPLY=($(compgen -W "all" -- "$cur"))
+      return # return early if we're still completing the 'current' command
+  fi
+
+
+  local cur="${COMP_WORDS[COMP_CWORD]}"
+  COMPREPLY=() #($(compgen -W "all" -- "$cur"))
+}
+
+
+_liquidctl_set_command ()
+{
+    local i=1 subcommand_index is_fan=-1
+
+  # find the sub command (either a fan or an led to set)
+  while [[ $i -lt $COMP_CWORD ]]; do
+      local s="${COMP_WORDS[i]}"
+      case "$s" in
+          fan[0-9] | fan | pump)
+              subcommand_index=$i
+              is_fan=1
+              break
+              ;;
+          led[0-9] | led | sync | ring | logo | external)
+              subcommand_index=$i
+              is_fan=0
+              break
+              ;;
+      esac
+      (( i++ ))
+  done
+
+  # check if it is a fan or an LED that is being set
+  if [[ "$is_fan" -eq "1" ]]
+  then
+      _liquidctl_set_fan
+  elif [[ "$is_fan" -eq "0" ]]
+  then
+      _liquidctl_set_led
+  else
+
+    # no trailing space here so that the fan number can be appended
+    compopt -o nospace
+    # possibly use some command here to get a list of all the possible channels from liquidctl
+    local cur="${COMP_WORDS[COMP_CWORD]}"
+    COMPREPLY=($(compgen -W "fan fan1 fan2 fan3 fan4 fan5 fan6 led led1 led2 led3 led4 led5 led6 pump sync ring logo external" -- "$cur"))
```

Comment:
I hadn't thought about simply suggesting all possible channel names, regardless of the particular device.  It's a small enough set, and it's better than nothing.  Cool!

Reply:
I really dont like that I hard coded all those values, I was originally thinking of added a new command to liquidctl (perhaps `--list-channels` and `--list-modes`) to get the values for specific devices, but the way the values are stored for each driver did not seem consistent at first glance and it would require a larger refactor then I really wanted to do for this.

Reply:
Yes, having a way to query the driver for what it (and the particular device that it's bounded to) support would be better, but also a lot of work.